rust/datafusion/src/datasource/parquet.rs

```diff
@@ -0,0 +1,571 @@
+            batch_size: 64 * 1024,
+            current_row_group: None,
+            column_readers: vec![],
+        })
+    }
+
+    fn load_next_row_group(&mut self) -> Result<()> {
+        if self.row_group_index < self.reader.num_row_groups() {
+            let reader = self.reader.get_row_group(self.row_group_index)?;
+
+            self.column_readers = Vec::with_capacity(self.projection.len());
+
+            for i in 0..self.projection.len() {
+                self.column_readers
+                    .push(reader.get_column_reader(self.projection[i])?);
+            }
+
+            self.current_row_group = Some(reader);
+            self.row_group_index += 1;
+
+            Ok(())
+        } else {
+            Err(ExecutionError::General(
+                "Attempt to read past final row group".to_string(),
+            ))
+        }
+    }
+
+    fn load_batch(&mut self) -> Result<Option<RecordBatch>> {
+        match &self.current_row_group {
+            Some(reader) => {
+                let mut batch: Vec<Arc<Array>> = Vec::with_capacity(reader.num_columns());
+                for i in 0..self.column_readers.len() {
+                    let is_nullable = self.schema().field(i).is_nullable();
+                    let array: Arc<Array> = match self.column_readers[i] {
+                        ColumnReader::BoolColumnReader(ref mut r) => {
+                            ArrowReader::<BooleanType>::read(
+                                r,
+                                self.batch_size,
+                                is_nullable,
+                            )?
+                        }
+                        ColumnReader::Int32ColumnReader(ref mut r) => {
+                            ArrowReader::<Int32Type>::read(
+                                r,
+                                self.batch_size,
+                                is_nullable,
+                            )?
+                        }
+                        ColumnReader::Int64ColumnReader(ref mut r) => {
+                            ArrowReader::<Int64Type>::read(
+                                r,
+                                self.batch_size,
+                                is_nullable,
+                            )?
+                        }
+                        ColumnReader::Int96ColumnReader(ref mut r) => {
+                            let mut read_buffer: Vec<Int96> =
+                                vec![Int96::new(); self.batch_size];
+
+                            let mut def_levels: Vec<i16> = vec![0; self.batch_size];
+                            let (_, levels_read) = r.read_batch(
+                                self.batch_size,
+                                Some(&mut def_levels),
+                                None,
+                                &mut read_buffer,
+                            )?;
+
+                            let mut builder =
+                                TimestampNanosecondBuilder::new(levels_read);
```

Comment:
Sorry about the back and forth @andygrove, the builder should be `TimestampMillisecondBuilder` in this instance as `convert_int96_timestamp` still returns milliseconds. Alternatively we could change `seconds * MILLIS_PER_SECOND + nanoseconds / 1_000_000` to `seconds * MILLIS_PER_SECOND * 1_000_000 + nanoseconds` to keep nano precision (which I prefer).

The original JIRA ticket that deprecated INT96 (https://issues.apache.org/jira/browse/PARQUET-323) based it on the reason that nanosecond precision is 'rarely a real requirement'.

Reply:
Good catch. Thanks. Since the schema converter says that INT96 is timestamp in nanoseconds (just like the C++ impl) we should convert to nanoseconds, so I updated as you suggested.